cmd/containerboot/main.go

```diff
@@ -0,0 +1,401 @@
+//                    store tailscaled state.
+//  - TS_SOCKS5_SERVER: the address on which to listen for SOCKS5
+//                      proxying into the tailnet.
+//  - TS_OUTBOUND_HTTP_PROXY_LISTEN: the address on which to listen
+//                                   for HTTP proxying into the tailnet.
+//  - TS_SOCKET: the path where the tailscaled local API socket should
+//               be created.
+//  - TS_AUTH_ONCE: if true, only attempt to log in if not already
+//                  logged in. If false (the default, for backwards
+//                  compatibility), forcibly log in every time the
+//                  container starts.
+//
+// When running on Kubernetes, TS_KUBE_SECRET takes precedence over
+// TS_STATE_DIR. Additionally, if TS_AUTH_KEY is not provided and the
+// TS_KUBE_SECRET contains an "authkey" field, that key is used.
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"net/netip"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+	"tailscale.com/client/tailscale"
+)
+
+func main() {
+	tailscale.I_Acknowledge_This_API_Is_Unstable = true
+
+	cfg := &settings{
+		AuthKey:         defaultEnv("TS_AUTH_KEY", ""),
+		Routes:          defaultEnv("TS_ROUTES", ""),
+		ProxyTo:         defaultEnv("TS_DEST_IP", ""),
+		DaemonExtraArgs: defaultEnv("TS_TAILSCALED_EXTRA_ARGS", ""),
+		ExtraArgs:       defaultEnv("TS_EXTRA_ARGS", ""),
+		InKubernetes:    os.Getenv("KUBERNETES_SERVICE_HOST") != "",
+		UserspaceMode:   defaultBool("TS_USERSPACE", true),
+		StateDir:        defaultEnv("TS_STATE_DIR", ""),
+		AcceptDNS:       defaultBool("TS_ACCEPT_DNS", false),
+		KubeSecret:      defaultEnv("TS_KUBE_SECRET", "tailscale"),
+		SOCKSProxyAddr:  defaultEnv("TS_SOCKS5_SERVER", ""),
+		HTTPProxyAddr:   defaultEnv("TS_OUTBOUND_HTTP_PROXY_LISTEN", ""),
+		Socket:          defaultEnv("TS_SOCKET", "/tmp/tailscaled.sock"),
+		AuthOnce:        defaultBool("TS_AUTH_ONCE", false),
+	}
+
+	if cfg.ProxyTo != "" && cfg.UserspaceMode {
+		log.Fatal("TS_DEST_IP is not supported with TS_USERSPACE")
+	}
+	if cfg.KubeSecret != "" && !cfg.InKubernetes {
+		log.Fatal("Cannot use TS_KUBE_SECRET outside of Kubernetes")
+	}
+
+	if !cfg.UserspaceMode {
+		if err := ensureTunFile(); err != nil {
+			log.Fatalf("Unable to create tuntap device file: %v", err)
+		}
+	}
+	if cfg.ProxyTo != "" || cfg.Routes != "" {
+		if err := ensureIPForwarding(); err != nil {
+			log.Printf("Failed to enable IP forwarding: %v", err)
+			log.Printf("To run tailscale as a proxy or router container, IP forwarding must be enabled.")
+			if cfg.InKubernetes {
+				log.Fatalf("You can either set the sysctls as a privileged initContainer, or run the tailscale container with privileged=true.")
+			} else {
+				log.Fatalf("You can fix this by running the container with privileged=true, or the equivalent in your container runtime that permits access to sysctls.")
+			}
+		}
+	}
+
```

Comment:
For lack of a better place to put it: if this is PID1 do you need to reap child processes? Something like this?
https://go.dev/play/p/CoDZcQ8uZys

Reply:
Switched the wait for just tailscaled to be a general PID reaper that also exits when the tailscaled PID exits.

Reply:
Oh, that's much cleaner :+1: